src/core/lib/layer-manager.js

```diff
@@ -583,8 +580,8 @@ export default class LayerManager {
           this._initializeLayer(newLayer);
           initLayerInSeer(newLayer); // Initializes layer in seer chrome extension (if connected)
         } else {
-          this._transferLayerState(oldLayer, newLayer);
-          this._updateLayer(newLayer);
+          this._markLayerMatched(oldLayer, newLayer);
+          this._updateLayer(newLayer, {transferStateFromLayer: oldLayer});
```

Comment:
I was thinking about getting rid of the odd-looking transferStateFromLayer parameter by just calling transferState here. 
It links to my wish to separate "dirtying of layers" from the actual update call.
This is the best approach I can think of right now.
Admittedly, it is kind of a "time limited" version of `oldProps`, so perhaps not perfect. But currently can't see how to do this without keeping a props copy around.
```
layer._transferState(oldLayer);
    // temporarily saves `layer.internalState.lastProps` if not already set.
layer._update(); 
     // if `layer.internalState.lastProps` is set, calls `layer.diffProps`. 
     // Calls `layer.updateState` with any changeFlags and `oldProps` set to `lastProps || this.props`.
     // Clears `layer.internalState.lastProps`;
```

Reply:
Well, I believe `changeFlags` are stored on `internalState` so in that sense, the difference is small.

A problem is that `changeFlags` is exposed to the app through the lifecycle methods so adding internal state to it increases exposure to internal state which one would prefer to avoid.

We could store the entire `updateParams` on `internalState` though I am not sure it buys us much, other than being able to stuff `oldProps` one level down.

I actually think that there is a big difference between storing `oldProps` on `Layer`, of which there will be an infinite succession, and storing them on `internalState` of which there will always only be one. It is a much more normal construction to reason about.